src/pip/_internal/cli/req_command.py

```diff
@@ -75,6 +76,37 @@ def _get_index_urls(cls, options):
         # Return None rather than an empty list
         return index_urls or None
 
+    @classmethod
+    def _get_headers(cls, options):
+        # type: (Values) -> Optional[Dict[str, str]]
+        """
+        Return a dict of extra HTTP request headers from user-provided options.
+        """
+        # Pull header inputs
+        header_inputs = getattr(options, 'headers', [])
+        if not header_inputs:
+            return None
+
+        # Refuse to set headers when multiple index URLs are set
+        index_urls = cls._get_index_urls(options)
+        if index_urls and len(index_urls) > 1:
+            logger.warning(
+                'Refusing to set -H / --header option(s) because multiple '
+                'index URLs are configured.',
+            )
+            return None
```

Comment:
I would suggest straight-up error here instead. It’d probably be a good idea to add a link to the discussion, so users can understand a rational instead of filing a new issue requesting the feature.

Reply:
I thought about that but I figured if the header was dropped and was critical for accessing the index (I'm really just thinking about using headers for auth stuff) then the command would surely fail downstream. But I'm also happy to make this change!